ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/util/SamplerSCollectionFunctions.scala

```diff
@@ -0,0 +1,253 @@
+      val (totalCount, keyCounts) = res
+      val totalDiff = (targetTotal - totalCount)/targetTotal
+      val keyDiffs = keyTargets.keySet.map(k =>
+          k -> (keyTargets(k) - keyCounts.getOrElse(k, 0L))/keyTargets(k)).toMap
+
+      (totalDiff, keyDiffs)
+    }.toSCollection
+  }
+
+  private[samplers] def buildUniformDiffs[T: ClassTag, U: ClassTag](s: SCollection[T],
+                                                                    sampled: SCollection[(U, T)],
+                                                                    keyFn: T => U,
+                                                                    prob: Double,
+                                                                    popPerKey: SideInput[Double])
+  : SCollection[(Double, Map[U, Double])] = {
+    //TODO: Error out if difference is too high
+    sampled.keys.map(k => (1L, Map[U, Long](k -> 1L))).sum
+      .withSideInputs(popPerKey).map{ case (res, sic) =>
+        val pop = sic(popPerKey)
+        val (totalCount, keyCounts) = res
+        val totalDiff = ((pop * keyCounts.size) - totalCount)/(pop * keyCounts.size)
+        val keyDiffs = keyCounts.keySet.map(k => k -> (pop - keyCounts.getOrElse(k, 0L))/pop).toMap
+
+        (totalDiff, keyDiffs)
+    }.toSCollection
+  }
+
+  private[samplers] def uniformParams[T: ClassTag, U: ClassTag](s: SCollection[T],
+                                                                keyFn: T => U,
+                                                                prob: Double)
+  : (SideInput[Double], SCollection[(U, Double)]) = {
+    val keyed = s.keyBy(keyFn)
+    val keys = keyed.keys.distinct
+    val keyCount = keys.count.asSingletonSideInput
+    val totalRecords = s.count
+    val populationPerKey: SideInput[Double] = totalRecords.withSideInputs(keyCount)
+      .map{case (c, sic) => (c * prob)/sic(keyCount)}.toSCollection.asSingletonSideInput
+    val probPerKey = keyed.countByKey.withSideInputs(populationPerKey).map {
+      case ((k, c), sic) => (k, sic(populationPerKey)/c) }.toSCollection
+    (populationPerKey, probPerKey)
+  }
+
+  private def filterByThreshold[T: ClassTag, U: ClassTag](s: SCollection[(U, (T, Double))],
+                                                           thresholdByKey: SCollection[(U, Double)])
+  : SCollection[(U, T)] = {
+    s.hashJoin(thresholdByKey)
+      .filter{case (_, ((_, d), t)) => d <= t}
+      .map{case (k, ((v, _), _)) => (k, v)}
+  }
+
+  private def stratifiedThresholdByKey[T: ClassTag, U: ClassTag](
+      s: SCollection[(U, (T, Double))],
+      countsByKey: SCollection[(U, (Long, Long))],
+      targetByKey: SCollection[(U, Long)],
+      variance: Double,
+      prob: Double)
+  : SCollection[(U, Double)] = {
+    s.map{case (k, (_, d)) => (k, d)}
+      .filter{case (_, d) => d < (prob + variance) && d > (prob - variance)}
+      //TODO: Clean up magic number
+      .topByKey(1e8.toInt)(Ordering.by(identity[Double]).reverse)
```

Comment:
Right now we need to pass this as an argument to `Top.of` on construction. We can potentially modify Top to provide this per-key but for now this will probably work. For larger datasets this will cause issues if the size of `variance * 2 * population > 1e8` and `count(elements where d < prob + variance) < target` .

Possible interim solution can also involve passing in an arg to overwrite this if necessary, but seems too low level to provide in the CLI API.